framework/src/constraints/AutomaticMortarGeneration.C

```diff
@@ -0,0 +1,1187 @@
+template <typename T>
+T
+fe_lagrange_1D_shape(const Order order, const unsigned int i, const T & xi)
+{
+  switch (order)
+  {
+      // Lagrange linears
+    case FIRST:
+    {
+      libmesh_assert_less(i, 2);
+
+      switch (i)
+      {
+        case 0:
+          return .5 * (1. - xi);
+
+        case 1:
+          return .5 * (1. + xi);
+
+        default:
+          mooseError("Invalid shape function index i = ", i);
+      }
+    }
+
+      // Lagrange quadratics
+    case SECOND:
+    {
+      libmesh_assert_less(i, 3);
+
+      switch (i)
+      {
+        case 0:
+          return .5 * xi * (xi - 1.);
+
+        case 1:
+          return .5 * xi * (xi + 1);
+
+        case 2:
+          return (1. - xi * xi);
+
+        default:
+          mooseError("Invalid shape function index i = ", i);
+      }
+    }
+
+    default:
+      mooseError("Unsupported order");
+  }
+}
+
+template <typename T>
+T
+fe_lagrange_1D_shape_deriv(const Order order, const unsigned int i, const T & xi)
+{
+  switch (order)
+  {
+      // Lagrange linear shape function derivatives
+    case FIRST:
+    {
+      libmesh_assert_less(i, 2);
+
+      switch (i)
+      {
+        case 0:
+          return -.5;
+
+        case 1:
+          return .5;
+
+        default:
+          mooseError("Invalid shape function index i = ", i);
+      }
+    }
+
+      // Lagrange quadratic shape function derivatives
+    case SECOND:
+    {
+      libmesh_assert_less(i, 3);
+
+      switch (i)
+      {
+        case 0:
+          return xi - .5;
+
+        case 1:
+          return xi + .5;
+
+        case 2:
+          return -2. * xi;
+
+        default:
+          mooseError("Invalid shape function index i = ", i);
+      }
+    }
+
+    default:
+      mooseError("Unsupported order");
+  }
+}
+
+const std::string AutomaticMortarGeneration::system_name = "Nodal Normals";
+
+AutomaticMortarGeneration::AutomaticMortarGeneration(
+    MeshBase & mesh_in,
+    const std::pair<BoundaryID, BoundaryID> & boundary_key,
+    const std::pair<SubdomainID, SubdomainID> & subdomain_key)
+  : mesh(mesh_in),
+    mortar_segment_mesh(mesh_in.comm()),
+    h_max(0.),
+    _periodic(false),
+    _periodic_set_externally(false)
+{
+  master_slave_boundary_id_pairs.push_back(boundary_key);
+  master_requested_boundary_ids.insert(boundary_key.first);
+  slave_requested_boundary_ids.insert(boundary_key.second);
+  master_slave_subdomain_id_pairs.push_back(subdomain_key);
+  master_boundary_subdomain_ids.insert(subdomain_key.first);
+  slave_boundary_subdomain_ids.insert(subdomain_key.second);
+}
+
+void
+AutomaticMortarGeneration::buildNodeToElemMaps()
+{
+  if (slave_requested_boundary_ids.empty() || master_requested_boundary_ids.empty())
+    libmesh_error_msg(
+        "Must specify slave and master boundary ids before building node-to-elem maps.");
+
+  // Construct nodes_to_slave_elem_map
+  for (MeshBase::const_element_iterator el = mesh.elements_begin(), end_el = mesh.elements_end();
```

Comment:
I don't think we have a range based version, but you can use the `as_range` helper to make one here.